crates/bevy_reflect/src/tuple_struct.rs

```diff
@@ -43,6 +47,48 @@ pub trait TupleStruct: Reflect {
     fn clone_dynamic(&self) -> DynamicTupleStruct;
 }
 
+/// A container for compile-time tuple struct info
+#[derive(Clone, Debug)]
+pub struct TupleStructInfo {
+    id: TypeIdentity,
+    fields: Box<[UnnamedField]>,
+}
+
+impl TupleStructInfo {
+    /// Create a new [`TupleStructInfo`]
+    ///
+    /// # Arguments
+    ///
+    /// * `fields`: The fields of this struct in the order they are defined
+    ///
+    pub fn new<T: Reflect>(fields: &[UnnamedField]) -> Self {
+        Self {
+            id: TypeIdentity::of::<T>(),
+            fields: fields.to_vec().into_boxed_slice(),
+        }
+    }
+
+    /// The [`TypeIdentity`] of this struct
+    pub fn id(&self) -> &TypeIdentity {
```

Comment:
I think I'd prefer to make TypeIdentity an internal implementation detail (or maybe remove it entirely to reduce the amount of abstraction / monomorphization. Types already have `TypeIds` and I think `TypeIdentity` is too "conceptually close" for comfort. We should try to be as "rust-native" as possible and avoid new jargon.

I think each `XInfo` type should have its own `type_name()` and `type_id()` functions.  Types like `MapInfo` that need to return type info in a consolidated fashion could return `ValueInfo`.

Reply:
> I think I'd prefer to make TypeIdentity an internal implementation detail (or maybe remove it entirely to reduce the amount of abstraction / monomorphization. Types already have `TypeIds` and I think `TypeIdentity` is too "conceptually close" for comfort. We should try to be as "rust-native" as possible and avoid new jargon.

The reason we originally switched to the `TypeIdentity` was to reduce the amount of duplicate code/docs between these info structs. Specifically, it contains both the `type_name` and `type_id` methods, as well as a helper method, `is<T>`.

I don't have an issue going back on this— I think you make a good point about staying as Rusty as we can,— but just want to make sure the slight increase in duplicated code/docs is okay with you.

Reply:
Another question is if returning `ValueInfo` might be somewhat confusing. Would users incorrectly believe that the underlying type is _not_ a struct or tuple since it's not a `StructInfo` or `TupleInfo`?

Reply:
> but just want to make sure the slight increase in duplicated code/docs is okay with you.

Yup this is ok with me. We pay that slight duplication cost on the implementation side, but it simplifies the user-facing side, which feels worth it to me.

> Another question is if returning ValueInfo might be somewhat confusing. Would users incorrectly believe that the underlying type is not a struct or tuple since it's not a StructInfo or TupleInfo?

Hmm yeah thats a solid take. Value means something very specific in this context and we shouldn't muddy the waters.  In a world where TypeInfo is low-cost to produce, we should probably just do `fn key(&self) -> &TypeInfo`. But that level of recursiveness feels a bit nasty in this context. I guess that leaves us with:

1. Return TypeIdentity, but only in cases like this with "nested" types. Slightly less consistent, but the common case has my preferred api.
2. Return TypeIdentity in all cases. Consistent, but extra abstraction + nesting to get to arguably the most important info.
3. For types with "nested info", use the form `fn key_type_id(&self) -> TypeId` and `fn key_type_name(&self) -> &'static str`. This is consistent, doesn't introduce new concepts, and keeps relevant info "flat' and easy to access.

I'm pretty sure I prefer (3), even though accessing "nested" type info is a bit clunkier.

Reply:
Yeah I’m not a fan of 1. And if we want to remove `TypeIdentity` in favor of being more Rusty/flat, then I agree 3 sounds like the way to go.